File: Pima Indians Diabetes Prediction - A Comparative Machine Learning Approach/scripts/r/univariate_bivariate_multivariate_analysis.R
# Pima Indians Diabetes Dataset - Univariate, Bivariate, and Multivariate Analysis
# R Script for Comprehensive Analysis

# Load necessary libraries
library(dplyr)
library(tidyverse)
library(ggplot2)
library(GGally)
library(corrplot)
library(gridExtra)
library(stats)

# Set output directory
output_dir <- "../../outputs/"
dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

# Function to load data
load_data <- function() {
  data <- read.csv("../../data/pima-indians-diabetes.csv", skip = 9, header = FALSE)
  colnames(data) <- c(
    "Pregnancies",
    "Glucose",
    "BloodPressure",
    "SkinThickness",
    "Insulin",
    "BMI",
    "DiabetesPedigreeFunction",
    "Age",
    "Outcome"
  )
  data$Outcome <- as.factor(data$Outcome)
  levels(data$Outcome) <- c("No Diabetes", "Diabetes")
  return(data)
}

# Univariate Analysis
univariate_analysis <- function(data, output_dir) {
  cat(paste(rep("=", 80), collapse = ""), "\n")
  cat("UNIVARIATE ANALYSIS\n")
  cat(paste(rep("=", 80), collapse = ""), "\n\n")
  
  numeric_features <- c("Pregnancies", "Glucose", "BloodPressure", "SkinThickness",
                       "Insulin", "BMI", "DiabetesPedigreeFunction", "Age")
  
  # Summary statistics
  univariate_stats <- data.frame(
    Feature = numeric_features,
    Mean = sapply(numeric_features, function(x) mean(data[[x]], na.rm = TRUE)),
    Median = sapply(numeric_features, function(x) median(data[[x]], na.rm = TRUE)),
    Std = sapply(numeric_features, function(x) sd(data[[x]], na.rm = TRUE)),
    Variance = sapply(numeric_features, function(x) var(data[[x]], na.rm = TRUE)),
    Skewness = sapply(numeric_features, function(x) {
      mean_val <- mean(data[[x]], na.rm = TRUE)
      sd_val <- sd(data[[x]], na.rm = TRUE)
      n <- sum(!is.na(data[[x]]))
      sum((data[[x]] - mean_val)^3, na.rm = TRUE) / (n * sd_val^3)
    }),
    Kurtosis = sapply(numeric_features, function(x) {
      mean_val <- mean(data[[x]], na.rm = TRUE)
      sd_val <- sd(data[[x]], na.rm = TRUE)
      n <- sum(!is.na(data[[x]]))
      sum((data[[x]] - mean_val)^4, na.rm = TRUE) / (n * sd_val^4) - 3
    }),
    Min = sapply(numeric_features, function(x) min(data[[x]], na.rm = TRUE)),
    Max = sapply(numeric_features, function(x) max(data[[x]], na.rm = TRUE)),
    Q1 = sapply(numeric_features, function(x) quantile(data[[x]], 0.25, na.rm = TRUE)),
    Q3 = sapply(numeric_features, function(x) quantile(data[[x]], 0.75, na.rm = TRUE)),
    IQR = sapply(numeric_features, function(x) IQR(data[[x]], na.rm = TRUE))
  )
  
  cat("Univariate Statistics:\n")
  print(univariate_stats, digits = 3)
  
  # Distribution plots
  plots <- list()
  for (i in 1:length(numeric_features)) {
    feature <- numeric_features[i]
    plots[[i]] <- ggplot(data, aes_string(x = feature)) +
      geom_histogram(bins = 30, fill = "steelblue", alpha = 0.7, color = "black") +
      geom_vline(aes_string(xintercept = mean(data[[feature]], na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
      geom_vline(aes_string(xintercept = median(data[[feature]], na.rm = TRUE)),
                 color = "green", linetype = "dashed", size = 1) +
      labs(title = paste("Distribution of", feature),
           x = feature, y = "Frequency") +
      theme_minimal()
  }
  
  pdf(paste0(output_dir, "univariate_distributions.pdf"), width = 20, height = 10)
  do.call(grid.arrange, c(plots, ncol = 4))
  dev.off()
  
  # Box plots
  plots <- list()
  for (i in 1:length(numeric_features)) {
    feature <- numeric_features[i]
    plots[[i]] <- ggplot(data, aes_string(y = feature)) +
      geom_boxplot(fill = "lightblue", alpha = 0.7) +
      labs(title = paste("Box Plot of", feature),
           y = feature) +
      theme_minimal()
  }
  
  pdf(paste0(output_dir, "univariate_boxplots.pdf"), width = 20, height = 10)
  do.call(grid.arrange, c(plots, ncol = 4))
  dev.off()
  
  # Normality tests
  cat("\n", paste(rep("=", 80), collapse = ""), "\n")
  cat("NORMALITY TESTS (Shapiro-Wilk Test)\n")
  cat(paste(rep("=", 80), collapse = ""), "\n\n")
  
  normality_results <- list()
  for (feature in numeric_features) {
    sample_size <- min(5000, nrow(data))
    sample_data <- sample(data[[feature]], sample_size)
    test_result <- shapiro.test(sample_data)
    normality_results[[feature]] <- data.frame(
      Feature = feature,
      W_statistic = test_result$statistic,
      P_value = test_result$p.value,
      Normal = ifelse(test_result$p.value > 0.05, "Yes", "No")
    )
  }
  
  normality_df <- do.call(rbind, normality_results)
  print(normality_df)
  
  return(list(univariate_stats = univariate_stats, normality_df = normality_df))
}

# Bivariate Analysis
bivariate_analysis <- function(data, output_dir) {
  cat("\n", paste(rep("=", 80), collapse = ""), "\n")
  cat("BIVARIATE ANALYSIS\n")
  cat(paste(rep("=", 80), collapse = ""), "\n\n")
  
  numeric_features <- c("Pregnancies", "Glucose", "BloodPressure", "SkinThickness",
                       "Insulin", "BMI", "DiabetesPedigreeFunction", "Age")
  
  # Feature vs Outcome box plots
  plots <- list()
  for (i in 1:length(numeric_features)) {
    feature <- numeric_features[i]
    plots[[i]] <- ggplot(data, aes_string(x = "Outcome", y = feature, fill = "Outcome")) +
      geom_boxplot(alpha = 0.7) +
      scale_fill_manual(values = c("skyblue", "salmon")) +
      labs(title = paste(feature, "by Outcome"),
           x = "Outcome", y = feature) +
      theme_minimal() +
      theme(legend.position = "none")
  }
  
  pdf(paste0(output_dir, "bivariate_boxplots.pdf"), width = 20, height = 10)
  do.call(grid.arrange, c(plots, ncol = 4))
  dev.off()
  
  # Density plots by outcome
  plots <- list()
  for (i in 1:length(numeric_features)) {
    feature <- numeric_features[i]
    plots[[i]] <- ggplot(data, aes_string(x = feature, fill = "Outcome")) +
      geom_density(alpha = 0.7) +
      scale_fill_manual(values = c("skyblue", "salmon"),
                       labels = c("No Diabetes", "Diabetes")) +
      labs(title = paste(feature, "Distribution by Outcome"),
           x = feature, y = "Density") +
      theme_minimal()
  }
  
  pdf(paste0(output_dir, "bivariate_densities.pdf"), width = 20, height = 10)
  do.call(grid.arrange, c(plots, ncol = 4))
  dev.off()
  
  # Correlation analysis
  data_numeric <- data
  data_numeric$Outcome <- as.numeric(data_numeric$Outcome) - 1
  
  correlation_matrix <- cor(data_numeric[, c(numeric_features, "Outcome")], use = "complete.obs")
  outcome_correlations <- correlation_matrix[, "Outcome"]
  outcome_correlations <- outcome_correlations[names(outcome_correlations) != "Outcome"]
  outcome_correlations <- sort(outcome_correlations, decreasing = TRUE)
  
  cat("Correlation with Outcome:\n")
  print(outcome_correlations, digits = 3)
  
  # Visualize correlations
  correlation_df <- data.frame(
    Feature = names(outcome_correlations),
    Correlation = as.numeric(outcome_correlations)
  )
  
  png(paste0(output_dir, "bivariate_correlations.png"), width = 800, height = 600)
  ggplot(correlation_df, aes(x = reorder(Feature, Correlation), y = Correlation)) +
    geom_bar(stat = "identity", fill = "steelblue", alpha = 0.7) +
    coord_flip() +
    labs(title = "Feature Correlations with Outcome",
         x = "Feature", y = "Correlation Coefficient") +
    theme_minimal()
  dev.off()
  
  # Statistical tests (T-tests)
  cat("\n", paste(rep("=", 80), collapse = ""), "\n")
  cat("STATISTICAL TESTS (Independent T-tests)\n")
  cat(paste(rep("=", 80), collapse = ""), "\n\n")
  
  diabetic <- data[data$Outcome == "Diabetes", ]
  non_diabetic <- data[data$Outcome == "No Diabetes", ]
  
  ttest_results <- list()
  for (feature in numeric_features) {
    test_result <- t.test(diabetic[[feature]], non_diabetic[[feature]])
    ttest_results[[feature]] <- data.frame(
      Feature = feature,
      T_statistic = test_result$statistic,
      P_value = test_result$p.value,
      Significant = ifelse(test_result$p.value < 0.05, "Yes", "No"),
      Diabetic_Mean = mean(diabetic[[feature]], na.rm = TRUE),
      Non_Diabetic_Mean = mean(non_diabetic[[feature]], na.rm = TRUE),
      Difference = mean(diabetic[[feature]], na.rm = TRUE) - mean(non_diabetic[[feature]], na.rm = TRUE)
    )
  }
  
  ttest_df <- do.call(rbind, ttest_results)
  print(ttest_df, digits = 4)
  
  return(list(correlations = outcome_correlations, ttest_df = ttest_df))
}

# Multivariate Analysis
multivariate_analysis <- function(data, output_dir) {
  cat("\n", paste(rep("=", 80), collapse = ""), "\n")
  cat("MULTIVARIATE ANALYSIS\n")
  cat(paste(rep("=", 80), collapse = ""), "\n\n")
  
  numeric_features <- c("Pregnancies", "Glucose", "BloodPressure", "SkinThickness",
                       "Insulin", "BMI", "DiabetesPedigreeFunction", "Age")
  
  # Correlation heatmap
  data_numeric <- data
  data_numeric$Outcome <- as.numeric(data_numeric$Outcome) - 1
  
  correlation_matrix <- cor(data_numeric[, c(numeric_features, "Outcome")], use = "complete.obs")
  
  png(paste0(output_dir, "multivariate_correlation_heatmap.png"), width = 1200, height = 1000)
  corrplot(correlation_matrix, method = "color", type = "upper", order = "hclust",
           tl.cex = 0.8, tl.col = "black", tl.srt = 45,
           addCoef.col = "black", number.cex = 0.7,
           col = colorRampPalette(c("blue", "white", "red"))(200),
           title = "Multivariate Analysis: Feature Correlation Heatmap", mar = c(0,0,2,0))
  dev.off()
  
  # PCA Analysis
  cat("\n", paste(rep("=", 80), collapse = ""), "\n")
  cat("PRINCIPAL COMPONENT ANALYSIS (PCA)\n")
  cat(paste(rep("=", 80), collapse = ""), "\n\n")
  
  # Prepare data (handle zeros)
  data_clean <- data_numeric[, numeric_features]
  data_clean[data_clean == 0] <- NA
  for (col in numeric_features) {
    data_clean[is.na(data_clean[[col]]), col] <- median(data_clean[[col]], na.rm = TRUE)
  }
  
  # Standardize
  data_scaled <- scale(data_clean)
  
  # Apply PCA
  pca <- prcomp(data_scaled, center = FALSE, scale. = FALSE)
  
  # Explained variance
  explained_variance <- pca$sdev^2 / sum(pca$sdev^2)
  cumulative_variance <- cumsum(explained_variance)
  
  cat("Explained Variance by Component:\n")
  for (i in 1:length(explained_variance)) {
    cat(sprintf("PC%d: %.4f (%.2f%%) - Cumulative: %.4f (%.2f%%)\n",
                i, explained_variance[i], explained_variance[i]*100,
                cumulative_variance[i], cumulative_variance[i]*100))
  }
  
  # Scree plot
  scree_df <- data.frame(
    PC = 1:length(explained_variance),
    Individual = explained_variance,
    Cumulative = cumulative_variance
  )
  
  png(paste0(output_dir, "multivariate_pca_scree.png"), width = 1000, height = 600)
  ggplot(scree_df, aes(x = PC)) +
    geom_bar(aes(y = Individual), stat = "identity", fill = "steelblue", alpha = 0.7) +
    geom_line(aes(y = Cumulative), color = "red", size = 1) +
    geom_point(aes(y = Cumulative), color = "red", size = 2) +
    labs(title = "PCA Scree Plot",
         x = "Principal Component", y = "Explained Variance Ratio") +
    theme_minimal()
  dev.off()
  
  # PCA visualization (2D)
  pca_2d <- data.frame(
    PC1 = pca$x[, 1],
    PC2 = pca$x[, 2],
    Outcome = data$Outcome
  )
  
  png(paste0(output_dir, "multivariate_pca_2d.png"), width = 1200, height = 800)
  ggplot(pca_2d, aes(x = PC1, y = PC2, color = Outcome)) +
    geom_point(alpha = 0.6, size = 2) +
    scale_color_manual(values = c("skyblue", "salmon"),
                      labels = c("No Diabetes", "Diabetes")) +
    labs(title = "PCA 2D Visualization",
         x = paste0("First Principal Component (Explained Variance: ",
                   round(explained_variance[1]*100, 2), "%)"),
         y = paste0("Second Principal Component (Explained Variance: ",
                   round(explained_variance[2]*100, 2), "%)")) +
    theme_minimal()
  dev.off()
  
  cat(sprintf("\nTotal Explained Variance (First 2 PCs): %.2f%%\n",
              sum(explained_variance[1:2])*100))
  
  # PCA component loadings
  pca_loadings <- data.frame(
    pca$rotation[, 1:2],
    Feature = numeric_features
  )
  colnames(pca_loadings) <- c("PC1", "PC2", "Feature")
  
  cat("\nPCA Component Loadings (First 2 Components):\n")
  print(pca_loadings, digits = 3)
  
  return(list(pca = pca, pca_loadings = pca_loadings))
}

# Main function
main <- function() {
  cat("Pima Indians Diabetes Dataset - Univariate, Bivariate, Multivariate Analysis\n")
  cat(paste(rep("=", 80), collapse = ""), "\n\n")
  
  # Load data
  data <- load_data()
  cat("Dataset loaded:", dim(data), "\n")
  cat("Features:", colnames(data), "\n\n")
  
  # Univariate analysis
  univariate_results <- univariate_analysis(data, output_dir)
  
  # Bivariate analysis
  bivariate_results <- bivariate_analysis(data, output_dir)
  
  # Multivariate analysis
  multivariate_results <- multivariate_analysis(data, output_dir)
  
  cat("\n", paste(rep("=", 80), collapse = ""), "\n")
  cat("ANALYSIS COMPLETE\n")
  cat(paste(rep("=", 80), collapse = ""), "\n\n")
  cat("Key Findings:\n")
  cat("1. Univariate: Features show varying distributions and normality\n")
  cat("2. Bivariate: Strong correlations between Glucose, BMI, Age and Outcome\n")
  cat("3. Multivariate: PCA reveals patterns in high-dimensional space\n")
  cat(paste(rep("=", 80), collapse = ""), "\n")
}

# Run main function
if (!interactive()) {
  main()
}

File: Pima Indians Diabetes Prediction - A Comparative Machine Learning Approach/scripts/python/univariate_bivariate_multivariate_analysis.py
"""
Pima Indians Diabetes Dataset - Univariate, Bivariate, and Multivariate Analysis
Python Script for Comprehensive Analysis
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
import warnings
warnings.filterwarnings('ignore')

# Set style
sns.set_style("whitegrid")
plt.rcParams['figure.figsize'] = (14, 8)

def load_data():
    """Load and prepare the dataset"""
    df = pd.read_csv('../../data/pima-indians-diabetes.csv', skiprows=9, header=None)
    columns = [
        'Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness',
        'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'
    ]
    df.columns = columns
    return df

def univariate_analysis(df, output_dir='../../outputs/'):
    """Perform univariate analysis on each feature"""
    print("=" * 80)
    print("UNIVARIATE ANALYSIS")
    print("=" * 80)
    
    numeric_features = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness',
                       'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age']
    
    # Summary statistics
    univariate_stats = pd.DataFrame({
        'Feature': numeric_features,
        'Mean': [df[col].mean() for col in numeric_features],
        'Median': [df[col].median() for col in numeric_features],
        'Std': [df[col].std() for col in numeric_features],
        'Variance': [df[col].var() for col in numeric_features],
        'Skewness': [df[col].skew() for col in numeric_features],
        'Kurtosis': [df[col].kurtosis() for col in numeric_features],
        'Min': [df[col].min() for col in numeric_features],
        'Max': [df[col].max() for col in numeric_features],
        'Q1': [df[col].quantile(0.25) for col in numeric_features],
        'Q3': [df[col].quantile(0.75) for col in numeric_features],
        'IQR': [df[col].quantile(0.75) - df[col].quantile(0.25) for col in numeric_features]
    })
    
    print("\nUnivariate Statistics:")
    print(univariate_stats)
    
    # Distribution plots
    fig, axes = plt.subplots(2, 4, figsize=(20, 10))
    axes = axes.ravel()
    
    for idx, feature in enumerate(numeric_features):
        axes[idx].hist(df[feature], bins=30, edgecolor='black', alpha=0.7, color='steelblue')
        axes[idx].axvline(df[feature].mean(), color='red', linestyle='--', label='Mean', linewidth=2)
        axes[idx].axvline(df[feature].median(), color='green', linestyle='--', label='Median', linewidth=2)
        axes[idx].set_title(f'Distribution of {feature}', fontsize=12, fontweight='bold')
        axes[idx].set_xlabel(feature)
        axes[idx].set_ylabel('Frequency')
        axes[idx].legend()
        axes[idx].grid(alpha=0.3)
    
    plt.suptitle('Univariate Analysis: Feature Distributions', fontsize=16, fontweight='bold', y=1.02)
    plt.tight_layout()
    plt.savefig(f'{output_dir}univariate_distributions.png', dpi=300, bbox_inches='tight')
    plt.show()
    
    # Box plots
    fig, axes = plt.subplots(2, 4, figsize=(20, 10))
    axes = axes.ravel()
    
    for idx, feature in enumerate(numeric_features):
        bp = axes[idx].boxplot(df[feature], vert=True, patch_artist=True)
        bp['boxes'][0].set_facecolor('lightblue')
        axes[idx].set_title(f'Box Plot of {feature}', fontsize=12, fontweight='bold')
        axes[idx].set_ylabel(feature)
        axes[idx].grid(alpha=0.3, axis='y')
    
    plt.suptitle('Univariate Analysis: Box Plots', fontsize=16, fontweight='bold', y=1.02)
    plt.tight_layout()
    plt.savefig(f'{output_dir}univariate_boxplots.png', dpi=300, bbox_inches='tight')
    plt.show()
    
    # Normality tests
    print("\n" + "=" * 80)
    print("NORMALITY TESTS (Shapiro-Wilk Test)")
    print("=" * 80)
    
    normality_results = []
    for feature in numeric_features:
        sample = df[feature].sample(min(5000, len(df)), random_state=42)
        stat, p_value = stats.shapiro(sample)
        normality_results.append({
            'Feature': feature,
            'W-statistic': stat,
            'P-value': p_value,
            'Normal': 'Yes' if p_value > 0.05 else 'No'
        })
    
    normality_df = pd.DataFrame(normality_results)
    print(normality_df)
    
    return univariate_stats, normality_df

def bivariate_analysis(df, output_dir='../../outputs/'):
    """Perform bivariate analysis between features and outcome"""
    print("\n" + "=" * 80)
    print("BIVARIATE ANALYSIS")
    print("=" * 80)
    
    numeric_features = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness',
                       'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age']
    
    # Feature vs Outcome box plots
    fig, axes = plt.subplots(2, 4, figsize=(20, 10))
    axes = axes.ravel()
    
    for idx, feature in enumerate(numeric_features):
        df.boxplot(column=feature, by='Outcome', ax=axes[idx], patch_artist=True)
        axes[idx].set_title(f'{feature} by Outcome', fontsize=12, fontweight='bold')
        axes[idx].set_xlabel('Outcome (0=No Diabetes, 1=Diabetes)')
        axes[idx].set_ylabel(feature)
        axes[idx].grid(alpha=0.3, axis='y')
    
    plt.suptitle('Bivariate Analysis: Features by Outcome', fontsize=16, fontweight='bold', y=1.02)
    plt.tight_layout()
    plt.savefig(f'{output_dir}bivariate_boxplots.png', dpi=300, bbox_inches='tight')
    plt.show()
    
    # Density plots by outcome
    fig, axes = plt.subplots(2, 4, figsize=(20, 10))
    axes = axes.ravel()
    
    for idx, feature in enumerate(numeric_features):
        df[df['Outcome'] == 0][feature].plot(kind='density', ax=axes[idx], 
                                             label='No Diabetes', alpha=0.7, color='skyblue', linewidth=2)
        df[df['Outcome'] == 1][feature].plot(kind='density', ax=axes[idx], 
                                             label='Diabetes', alpha=0.7, color='salmon', linewidth=2)
        axes[idx].set_title(f'{feature} Distribution by Outcome', fontsize=12, fontweight='bold')
        axes[idx].set_xlabel(feature)
        axes[idx].set_ylabel('Density')
        axes[idx].legend()
        axes[idx].grid(alpha=0.3)
    
    plt.suptitle('Bivariate Analysis: Feature Distributions by Outcome', fontsize=16, fontweight='bold', y=1.02)
    plt.tight_layout()
    plt.savefig(f'{output_dir}bivariate_densities.png', dpi=300, bbox_inches='tight')
    plt.show()
    
    # Correlation analysis
    correlation_matrix = df[numeric_features + ['Outcome']].corr()
    outcome_correlations = correlation_matrix['Outcome'].drop('Outcome').sort_values(ascending=False)
    
    print("\nCorrelation with Outcome:")
    print(outcome_correlations)
    
    # Visualize correlations
    plt.figure(figsize=(10, 6))
    outcome_correlations.plot(kind='barh', color='steelblue')
    plt.title('Feature Correlations with Outcome', fontsize=14, fontweight='bold')
    plt.xlabel('Correlation Coefficient')
    plt.grid(alpha=0.3, axis='x')
    plt.tight_layout()
    plt.savefig(f'{output_dir}bivariate_correlations.png', dpi=300, bbox_inches='tight')
    plt.show()
    
    # Statistical tests (T-tests)
    print("\n" + "=" * 80)
    print("STATISTICAL TESTS (Independent T-tests)")
    print("=" * 80)
    
    diabetic = df[df['Outcome'] == 1]
    non_diabetic = df[df['Outcome'] == 0]
    
    ttest_results = []
    for feature in numeric_features:
        stat, p_value = stats.ttest_ind(diabetic[feature], non_diabetic[feature])
        ttest_results.append({
            'Feature': feature,
            'T-statistic': stat,
            'P-value': p_value,
            'Significant': 'Yes' if p_value < 0.05 else 'No',
            'Diabetic Mean': diabetic[feature].mean(),
            'Non-Diabetic Mean': non_diabetic[feature].mean(),
            'Difference': diabetic[feature].mean() - non_diabetic[feature].mean()
        })
    
    ttest_df = pd.DataFrame(ttest_results)
    print(ttest_df)
    
    return outcome_correlations, ttest_df

def multivariate_analysis(df, output_dir='../../outputs/'):
    """Perform multivariate analysis"""
    print("\n" + "=" * 80)
    print("MULTIVARIATE ANALYSIS")
    print("=" * 80)
    
    numeric_features = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness',
                       'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age']
    
    # Correlation heatmap
    correlation_matrix = df[numeric_features + ['Outcome']].corr()
    
    plt.figure(figsize=(12, 10))
    sns.heatmap(correlation_matrix, annot=True, fmt='.2f', cmap='coolwarm', 
                center=0, square=True, linewidths=1, cbar_kws={"shrink": 0.8},
                vmin=-1, vmax=1)
    plt.title('Multivariate Analysis: Feature Correlation Heatmap', fontsize=14, fontweight='bold')
    plt.tight_layout()
    plt.savefig(f'{output_dir}multivariate_correlation_heatmap.png', dpi=300, bbox_inches='tight')
    plt.show()
    
    # PCA Analysis
    print("\n" + "=" * 80)
    print("PRINCIPAL COMPONENT ANALYSIS (PCA)")
    print("=" * 80)
    
    # Prepare data (handle zeros)
    df_clean = df[numeric_features].copy()
    df_clean = df_clean.replace(0, np.nan)
    df_clean = df_clean.fillna(df_clean.median())
    
    # Standardize
    scaler = StandardScaler()
    X_scaled = scaler.fit_transform(df_clean)
    
    # Apply PCA
    pca = PCA(n_components=min(8, len(numeric_features)))
    X_pca = pca.fit_transform(X_scaled)
    
    # Explained variance
    explained_variance = pca.explained_variance_ratio_
    cumulative_variance = np.cumsum(explained_variance)
    
    print(f"Explained Variance by Component:")
    for i, (var, cum_var) in enumerate(zip(explained_variance, cumulative_variance)):
        print(f"PC{i+1}: {var:.4f} ({var*100:.2f}%) - Cumulative: {cum_var:.4f} ({cum_var*100:.2f}%)")
    
    # Scree plot
    plt.figure(figsize=(10, 6))
    plt.bar(range(1, len(explained_variance) + 1), explained_variance, alpha=0.7, color='steelblue', label='Individual')
    plt.plot(range(1, len(cumulative_variance) + 1), cumulative_variance, 'ro-', label='Cumulative')
    plt.xlabel('Principal Component')
    plt.ylabel('Explained Variance Ratio')
    plt.title('PCA Scree Plot', fontsize=14, fontweight='bold')
    plt.legend()
    plt.grid(alpha=0.3)
    plt.tight_layout()
    plt.savefig(f'{output_dir}multivariate_pca_scree.png', dpi=300, bbox_inches='tight')
    plt.show()
    
    # PCA visualization (2D)
    pca_2d = PCA(n_components=2)
    X_pca_2d = pca_2d.fit_transform(X_scaled)
    
    plt.figure(figsize=(12, 8))
    scatter = plt.scatter(X_pca_2d[:, 0], X_pca_2d[:, 1], c=df['Outcome'], 
                         cmap='viridis', alpha=0.6, edgecolors='black', s=50)
    plt.colorbar(scatter, label='Outcome')
    plt.xlabel(f'First Principal Component (Explained Variance: {pca_2d.explained_variance_ratio_[0]:.2%})')
    plt.ylabel(f'Second Principal Component (Explained Variance: {pca_2d.explained_variance_ratio_[1]:.2%})')
    plt.title('PCA 2D Visualization', fontsize=14, fontweight='bold')
    plt.grid(alpha=0.3)
    plt.tight_layout()
    plt.savefig(f'{output_dir}multivariate_pca_2d.png', dpi=300, bbox_inches='tight')
    plt.show()
    
    # PCA component loadings
    pca_loadings = pd.DataFrame(
        pca_2d.components_.T,
        columns=['PC1', 'PC2'],
        index=numeric_features
    )
    print("\nPCA Component Loadings (First 2 Components):")
    print(pca_loadings)
    
    # 3D scatter plot
    from mpl_toolkits.mplot3d import Axes3D
    
    top_features = ['Glucose', 'BMI', 'Age']
    fig = plt.figure(figsize=(14, 10))
    ax = fig.add_subplot(111, projection='3d')
    
    scatter = ax.scatter(df[top_features[0]], df[top_features[1]], df[top_features[2]], 
                        c=df['Outcome'], cmap='viridis', alpha=0.6, edgecolors='black', s=50)
    ax.set_xlabel(top_features[0])
    ax.set_ylabel(top_features[1])
    ax.set_zlabel(top_features[2])
    ax.set_title('3D Scatter Plot: Top 3 Features (Colored by Outcome)', fontsize=14, fontweight='bold')
    plt.colorbar(scatter, label='Outcome')
    plt.savefig(f'{output_dir}multivariate_3d_scatter.png', dpi=300, bbox_inches='tight')
    plt.show()
    
    return pca, pca_loadings

def main():
    """Main function to run all analyses"""
    print("Pima Indians Diabetes Dataset - Univariate, Bivariate, Multivariate Analysis")
    print("=" * 80)
    
    # Create output directory
    import os
    os.makedirs('../../outputs', exist_ok=True)
    
    # Load data
    df = load_data()
    print(f"\nDataset loaded: {df.shape}")
    print(f"Features: {list(df.columns)}")
    
    # Univariate analysis
    univariate_stats, normality_df = univariate_analysis(df)
    
    # Bivariate analysis
    correlations, ttest_df = bivariate_analysis(df)
    
    # Multivariate analysis
    pca_model, pca_loadings = multivariate_analysis(df)
    
    print("\n" + "=" * 80)
    print("ANALYSIS COMPLETE")
    print("=" * 80)
    print("\nKey Findings:")
    print("1. Univariate: Features show varying distributions and normality")
    print("2. Bivariate: Strong correlations between Glucose, BMI, Age and Outcome")
    print("3. Multivariate: PCA reveals patterns in high-dimensional space")
    print("=" * 80)

if __name__ == "__main__":
    main()


